File: library/java/flow/MinCostMaxFlow.java
package flow;

import java.util.ArrayList;
import java.util.PriorityQueue;
import static java.lang.Math.*;

public class MinCostMaxFlow {
	static final int INF = 1001001001;

	int minCostMaxFlow(V[] vs, V s, V t, int flow) {
		int res = 0;
		while (flow > 0) {
			for (V v : vs) v.min = INF;
			PriorityQueue<E> que = new PriorityQueue<E>();
			s.min = 0;
			que.offer(new E(s, 0, 0));
			while (!que.isEmpty()) {
				E crt = que.poll();
				if (crt.cost == crt.to.min) {
					for (E e : crt.to.es) {
						int tmp = crt.cost + e.cost + crt.to.h - e.to.h;
						if (e.cap > 0 && e.to.min > tmp) {
							e.to.min = tmp;
							e.to.prev = e;
							que.offer(new E(e.to, 0, e.to.min));
						}
					}
				}
			}
			if (t.min == INF) return INF;
			int d = flow;
			for (E e = t.prev; e != null; e = e.rev.to.prev) {
				d = min(d, e.cap);
			}
			for (E e = t.prev; e != null; e = e.rev.to.prev) {
				res += d * e.cost;
				e.cap -= d;
				e.rev.cap += d;
			}
			flow -= d;
			for (V v : vs)
				v.h += v.min;
		}
		return res;
	}

	class V {
		int id;
		ArrayList<E> es = new ArrayList<E>();
		E prev;
		int min, h;

		void add(V to, int cap, int cost) {
			E e = new E(to, cap, cost), rev = new E(this, 0, -cost);
			e.rev = rev;
			rev.rev = e;
			es.add(e);
			to.es.add(rev);
		}
	}

	class E implements Comparable<E> {
		V to;
		E rev;
		int cap, cost;

		E(V to, int cap, int cost) {
			this.to = to;
			this.cap = cap;
			this.cost = cost;
		}

		public int compareTo(E o) {
			return cost - o.cost;
		}
	}
}

File: library/cpp/flow/mincost_maxflow.cpp
const int MAX_V = 20000+5;
using CAP = int;
using COST = int;
const int INF = (int)1e9;

struct State {
  COST dist;
  int pos;
  bool operator < (const State &o) const {
    return dist > o.dist; // reverse order
  }
};
struct Edge{ int to; CAP cap; COST cost; int rev; };
vector<Edge> adj[MAX_V];
COST potential[MAX_V];
COST dist[MAX_V];
int prevv[MAX_V];
int preve[MAX_V];
void clearedge() { REP(i, MAX_V) adj[i].clear(); }
void addedge(int u, int v, CAP cap, COST cost) {
  adj[u].push_back( (Edge){v, cap, cost, (int)adj[v].size() + (u == v)} );
  adj[v].push_back( (Edge){u, 0, -cost, (int)adj[u].size() - 1} );
}

// sからtへ流量 flow を流す最小コストを返す。
// 流せない場合は INF を返す
COST mincost_maxflow(int s, int t, CAP flow) {
  vector<int> vs;
  REP(i, MAX_V) if (adj[i].size()) vs.push_back(i);
  COST res = 0;
  for (int v : vs) potential[v] = 0;
  while (flow > 0) {
    for (int v : vs) dist[v] = INF;
    priority_queue<State> q;
    q.push({0, s});
    dist[s] = 0;
    while (!q.empty()) {
      State crt = q.top(); q.pop();
      if (dist[crt.pos] < crt.dist) continue;
      REP(i, adj[crt.pos].size()) {
        const Edge&e = adj[crt.pos][i];
        COST n_dist = crt.dist + e.cost + potential[crt.pos] - potential[e.to];
        if (e.cap > 0 && dist[e.to] > n_dist) {
          dist[e.to] = n_dist;
          prevv[e.to] = crt.pos;
          preve[e.to] = i;
          q.push({dist[e.to], e.to});
        }
      }
    }
    if (dist[t] == INF) return INF;
    for (int v : vs) potential[v] += dist[v];
    CAP f = flow;
    for (int v = t; v != s; v = prevv[v]) {
      f = min(f, adj[prevv[v]][preve[v]].cap);
    }
    for (int v = t; v != s; v = prevv[v]) {
      Edge &e = adj[prevv[v]][preve[v]];
      e.cap -= f;
      adj[v][e.rev].cap += f;
    }
    flow -= f;
    res += f * potential[t];
  }
  return res;
}
